Added initial version of last-updated extension

maxberger opened the pull request:
See https://github.com/homieiot/convention/issues/191 for discussion.

maxberger committed:
Added initial version of last-updated extension

EPNW commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,73 @@
+## Homie Version
+This extension supports Homie `3.0.1` and `4.x`.
+
+## Extension Identifier
```
There is something wrong with your extension identifier.  Assuming you intended the ID `name.berger.name.last-updated`, which seems a bit odd to me, the  **$extensions entry** should be `name.berger.name.last-updated:1.0.0:[3.0.1;4.x]`. 

I would suggest using a more convenient ID in domain format, like [your country code].berger.last-updated, but thats up to you :)

maxberger replied:
Apologies, there was indeed a spelling mistake here. It should be name.berger.max in order to follow the convention of matching website-owner to id-owner.  (My personal website is https://max.berger.name/ ).

Of course, we could also use the homie identifier, but then this would seem like an "official" extension. Would that be preferred?

davidgraeff replied:
Very often the domain registrar is actually used to identify a purpose like "org" or "io" for public use / invented by public organisations or individuals vs "com" for commercial invented things or used purposes. That's how I see it used in the Java world at least. The reverse domain doesn't always actually exist.

Because the "name" TLD is not common I wouldn't use it here, as it is hardly identified as being a TLD. I suggest something like "org.berger.last-updated" or similar.

maxberger replied:
changed id again

maxberger committed:
Fix typo in ID

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,73 @@
+
+| Topic                                 | Description                                       | Payload type                       |
+|---------------------------------------|---------------------------------------------------|------------------------------------|
+| $last-updated        | Timestamp when the data was last updated from the backend / sensor              | integer: unix timestamp in UTC (seconds since 1/1/70) OR 0 for unknown |
```
```suggestion
| $last-updated        | Timestamp when the property value was last updated by the Homie device | integer: seconds that have passed since the Unix epoch (00:00:00 UTC on 1 January 1970) OR 0 for unknown |
```

to be more precise and use the official Homie terminology

ThomDietrich replied:
Additionally: Why is 0 allowed? The use case is unclear and the meaning ambiguous (it's literally just 00:00:00 UTC on 1 January 1970)

Edit: See my longer comment below.

Also: I'm reluctant whether or not I care about the slight ambiguity. Probably totally fine

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,73 @@
+|---------------------------------------|---------------------------------------------------|------------------------------------|
+| $last-updated        | Timestamp when the data was last updated from the backend / sensor              | integer: unix timestamp in UTC (seconds since 1/1/70) OR 0 for unknown |
+
+Device usage:
```
"Device"?

maxberger replied:
Rewritten to make it clearer

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,73 @@
+
+This means that the property pressure was last fecthed from its source at unix timestamp 1579965965 (2020 Jan 25, 16:26:05 UTC, or 2020 Jan 25, 17:26:05 CET)
+
+**Client considerations**
```
Who is this directed at? Please align "Client" to the Homie lingo

ThomDietrich replied:
Also add how to react to value 0 here. That would probably answer my question from above.

maxberger replied:
Added example, renamed to controller, added (hopefully) better explanation for the 0 value.

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,73 @@
+Device usage:
+* The $last-update **must** always be published immediately before a value update.
+* $last-update **must** be 0 or a positive integer
+* $last-update **must** be in UTC
```
Are these redundant?

maxberger replied:
If they are, I prefer to be overly-specific in this case.

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,73 @@
+* $last-update **must** be 0 or a positive integer
+* $last-update **must** be in UTC
+* If a device cannot get the actual timestamp for whatever reason (clock is skewed, ntp has not yet finished, etc.) it **may** publish 0
+* If a device has published a $last-update in the past and is now unable to determine the last-update timestamp, it **must** publish 0 at least once.
```
Is this different to unsetting the topic?

maxberger replied:
It is hard to "unset" a topic in MQTT; we could publish blank or " ", but that would then not be a valid integer. Using 0 has the advantage that controller which don't want to interpret the value specially will have a valid, transformable timestamp. 

Added this to the doc.

ThomDietrich replied:
Unsetting is technically possible. See "Delete": https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/

However, to determine that a Homie device uses the extension publishing any value to the topic (timestamp or 0) is the right way forward.

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
````diff
@@ -0,0 +1,73 @@
+homie/super-car/wheels/pressure/ → 32
+```
+
+This means that the property pressure was last fecthed from its source at unix timestamp 1579965965 (2020 Jan 25, 16:26:05 UTC, or 2020 Jan 25, 17:26:05 CET)
````
*fetched

maxberger replied:
fixed.

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,73 @@
+
+**Client considerations**
+
+Note that MQTT is not atomic, and does not guarantee any order. This means that even though both the value and the last-update timestamp are published together and in the oder above, the clients may receive these messages in any order, or one of them may be lost. Therefore, **$last-updated** is entirely **best-effort**. It **must** **not** be used for any critical operation.
```
comma after atomic?

Because of your argumentation I would actually not state any publishing order up top. If reception in a specific order can't be promised the controller can't act on it either.  Thinking about this some more I now understand one potential purpose of the 0 value. We could define in the extension that a controller has to expect an updated timestamp with every property value update as long as the $last-updated topic is set.

maxberger replied:
changed to*should* above.

ThomDietrich reviewed (changes requested):
Hey!

Some thoughts on your way the two topics are synchronized and what the value 0 stands for. Imho it would be good to give clearer instructions how to process data for the discovering client/controller.

Like the extension 👌 Best!

maxberger commented:
Sorry for the late reply, my life kept me busy. I am now back and trying to get this into good shape.

I do see a strong push for consistency from multiple side, and I would like to propose yet another alternative. 

Option 1: Keep the proposal as-is, addressing the comments above.

Option 2: Completely re-think the prosposal and provide both the value and the last-updated in the same property. So here is what we could do:
* The extension would be a more generic "value-with-metadata" extension.
* The original value would always be published as normal, but
* An additional "extended" value would be written. This value would follow a common format (probably JSON), and would contain additional meta information, which combines both the value AND the last-updated, and possible other things in the same update

Example:
* homie/super-car/wheels/pressure/ → 32
* homie/super-car/wheels/pressure/$ext → "{value: 32, last-updated: 123456}"

Of course, here we could run into the same consistency issues, but at least a client could look at either one of the two values. It would also eliminate the need for a "0" timestamp.

Pros:
* Extensible
* In it self consistent
* Writing JSON is cheap.

Cons:
* Duplication of values
* clients which would need the extended data would need to parse JSON

wdyt?

davidgraeff commented:
> Writing JSON is cheap.

Homie has deliberately not used JSON anywhere. For extensions this should be fine, especially if multiple values need to be kept in sync. It if is a fixed known amount of values, CSV could be an option as well though.

ThomDietrich commented:
Hey all, 
I see how option 1 is flawed (https://github.com/homieiot/convention/issues/191#issuecomment-580298368) but option 2 I would avoid as well.First of all as @davidgraeff said, we decided to stay clear of JSON and we should try to uphold this principle for consistency.
Furthermore, an extension for "value with metadata" feels too far fetched. Remember that one can always implement additional topics with arbitrary complexity in parallel to what is defined by Homie. The Homie convention is here to aid automatic discovery of devices. An extension to process a last-updated makes sense, some generic metadata does not so much.

Following my interpretation of use cases for this extension (@maxberger this should be part of the intro) I propose a **third option**: The extension defines two topics:
* `$last-updated` - epoch time as defined so far, e.g. "1579965965"
* `$last-value-updated` - A combination of value and epoch time, e.g. "12.5;1579965965"

This combination offers a dateime-only topic to directly get a timestamp for keep-alive use cases, and an atomic value-datetime pair for instrumentation/monitoring use cases. Justification for having both would be the ease of processing on both the client and controller side.

Does this cover all use cases? Does anyone dislike the redundancy?

maxberger committed:
Update extensions/documents/homie_last_updated_extension.md

Co-Authored-By: Thomas Dietrich <Thomas@Nurzen.de>

maxberger committed:
Added updates based on discussion

https://github.com/homieiot/convention/pull/192

maxberger commented:
Like the idea; added this to the extension as well.

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,134 @@
+## Abstract
+This extension adds best-effort last-updated timestamps to properties.
+
+Homie devices will publish a last-updated timestamp along with any property value changes. The last-update timestamp can be sent by devices or gateways that do not constantly pull data from the real back-end. In this case, they can report when they last actually updated the data, so that controllers can act accordingly, and maybe ignore old data, or display this information in an UI.
```
General style thing: Please break the whole article down to one sentence per line. The rendered output will be the same but this makes a few things easier.

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,134 @@
+## Abstract
+This extension adds best-effort last-updated timestamps to properties.
+
+Homie devices will publish a last-updated timestamp along with any property value changes. The last-update timestamp can be sent by devices or gateways that do not constantly pull data from the real back-end. In this case, they can report when they last actually updated the data, so that controllers can act accordingly, and maybe ignore old data, or display this information in an UI.
```
I would suggest to start with a rationale sentence or even section, something like

> In most Homie application scenarios it is implicitly assumed that a property update on MQTT directly follows the real world event, e.g., the press of a physical switch would trigger an MQTT message and a controller would react to said message and turn a light on.
>
> This extension addresses the special use case of a Homie device in which the timeliness of an event can't be guaranteed or in which the exact timestamp of a value is of importance.
> Examples for such use cases are
> * ...

ThomDietrich replied:
You sometimes write "last-update", sometimes "last-updated". I understand the incentive but this might be confusing.

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,134 @@
+
+Homie devices will publish a last-updated timestamp along with any property value changes. The last-update timestamp can be sent by devices or gateways that do not constantly pull data from the real back-end. In this case, they can report when they last actually updated the data, so that controllers can act accordingly, and maybe ignore old data, or display this information in an UI.
+
+A common use case for this is a gateway software which publishes device data in Homie format, but actually uses external sensors to get the actual data. Gateways may re-publish the values the gather from the backend, even if these values are older. Using the last-update timestamp, a controller is then able to distinguish between actual value updates, and value updates which are just a republicati ...
```
👍

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,134 @@
+The ID of this extension is `org.homie.ext.last-value-updated`.
+Therefore the **$extensions entry** is `org.homie.ext.last-value-updated:1.0.0:[3.0.1;4.x]`.
+
+A device may implement zero, one, or both extensions. The extension entries refer to the matching property attributes.
```
I suggest to add:

> A controller supporting the extension must support both topics

ThomDietrich replied:
I would not separate the extension into two.
It feels rather confusing and both topics are for the same purpose behind this single extension...

ThomDietrich commented on extensions/documents/homie_last_updated_extension.md:
```diff
@@ -0,0 +1,134 @@
+
+#### $last-updated
+
+The **$last-updated** property attribute is **optional**.
```
"Optional" renders the whole extension ambiguous.

Let's define the following: A Homie device may implement neither of the topics, one topic, or both topics. A Homie controller supporting the extension must support both topics.

With that the rule is simple and deterministic:
A Homie controller supporting the extension shall always evaluate `$last-value-updated` when defined, otherwise `$last-updated` and the property value.

If we agree so far, we can go one step further:
Both topics can be combined into one as the controller is able to parse which of the two formats is used and there is no case in which it actually makes sense for a Homie device to publish both formats.
Wdyt?

ThomDietrich commented:
There is one further design issue we need to discuss. @davidgraeff please add your two cents.

In the Homie convention dollar topics are used to provide metadata for controller-side feature discovery of a device. Your "last-updated" topic communicates data. This is clearly not in line with the convention. Therefore, I think we need to discuss one further change to the topic structure of this extension. Wdyt?

mjcumming commented:
@maxberger to get around this problem I created a jython script 

https://community.openhab.org/t/asynchronous-events-to-synchronous-event/95198

mjcumming commented:
@maxberger  and @ThomDietrich 

see http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718105 from @Thalhammer  in [#200](url)  There is ordering of MQTT messages of the same QoS.

Would adding a property attribute $lastupated and specifying that this attribute must be updated after the value has been set solve this is?

Thalhammer commented:
> @maxberger and @ThomDietrich
> 
> see http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718105 from @Thalhammer in [#200](url) There is ordering of MQTT messages of the same QoS.
> 
> Would adding a property attribute $lastupated and specifying that this attribute must be updated after the value has been set solve this is?

While acording to the rfc MQTT *should* be ordered withing one qos value, I don't think we should rely to badly on it. Not all brokers might follow the spec that closely.

maxberger commented:
I no longer have the time to follow up with this.
